src/Services/StateManager.ts

```diff
@@ -0,0 +1,140 @@
+import { ActionType, MovieDbStateAction } from "../types/movieActions";
+import { Movie, MovieDbState, NewMovie, SortOrder, UpdateActivity } from "../types/movieModels";
+
+type SortHandler = (m1: Movie, m2: Movie) => number;
+
+type SortOrderHandlers = Record<SortOrder, SortHandler>;
+
+const sortOrderHandlers: SortOrderHandlers = {
+    [SortOrder.ByNameAsc]: (m1: Movie, m2: Movie) => m1.title.toLocaleLowerCase().localeCompare(m2.title.toLocaleLowerCase()),
+    [SortOrder.ByNameDesc]: (m1: Movie, m2: Movie) => m2.title.toLocaleLowerCase().localeCompare(m1.title.toLocaleLowerCase()),
+    [SortOrder.ByReleaseAsc]: (m1: Movie, m2: Movie) => m1.year.localeCompare(m2.year),
+    [SortOrder.ByReleaseDesc]: (m1: Movie, m2: Movie) => m2.year.localeCompare(m1.year),
+};
+
+export default class MovieDbStateManager {
+
+    static initState(initList: ReadonlyArray<Movie>): Readonly<MovieDbState> {
+        const sortOrder = SortOrder.ByReleaseAsc;
+        const moviesList = [...initList].sort(sortOrderHandlers[sortOrder]);
+        const genres = MovieDbStateManager.getGenresFromMovies(moviesList);
+        return { moviesList, sortOrder, genres, filteredMoviesList: moviesList };
+    }
+
+    static reducer(state: Readonly<MovieDbState>, action: Readonly<MovieDbStateAction>): Readonly<MovieDbState> {
+        switch (action.type) {
+            case ActionType.AddMovie:
+                return MovieDbStateManager.handleAddMovie(state, action.payload);
+            case ActionType.DeleteMovie:
+                return MovieDbStateManager.handleDeleteMovie(state, action.payload);
+            case ActionType.EditMovie:
+                return MovieDbStateManager.handleEditMovie(state, action.payload);
+            case ActionType.SelectMovie:
+                return MovieDbStateManager.handleSelectMovie(state, action.payload);
+            case ActionType.SortMovies:
+                return MovieDbStateManager.handleSortMovies(state, action.payload);
+            case ActionType.ShowAddMovie:
+                return MovieDbStateManager.handleShowHideAction(state, UpdateActivity.addActivity);
+            case ActionType.ShowEditMovie:
+                return MovieDbStateManager.handleShowHideAction(state, UpdateActivity.editActivity, action.payload);
+            case ActionType.ShowDeleteMovie:
+                return MovieDbStateManager.handleShowHideAction(state, UpdateActivity.deleteActivity, action.payload);
+            case ActionType.HideMovieUpdate:
+                return MovieDbStateManager.handleShowHideAction(state);
+            case ActionType.ActiveGenreChange:
+                return MovieDbStateManager.handleActiveGenreChange(state, action.payload);
+            default:
+                const _exhaustiveCheck: never = action;
+                return _exhaustiveCheck;
```

Comment:
на мій погляд буде більш зрозуміло, якщо ти тут явно викинеш помилку
```suggestion
                throw new Error(`Unhandled action ${action.type}`)
```